verdigado_attendance/models/hr_attendance.py

```diff
@@ -9,3 +9,83 @@
 
 class HrAttendance(models.Model):
     _inherit = "hr.attendance"
+
+    def _update_overtime(self, employee_attendance_dates=None):
+        """Compute overtime from theoretical time"""
+        report = self.env["hr.attendance.theoretical.time.report"]
+        for this in self:
+            day_start_utc, date = self._get_day_start_and_day(
+                this.employee_id, this.check_in
+            )
+            overtime = (
+                self.env["hr.attendance.overtime"]
+                .search(
+                    [
+                        ("employee_id", "=", this.employee_id.id),
+                        ("date", "=", date),
+                        ("adjustment", "=", False),
+                    ]
+                )
+                .sudo()
+            )
+            if not overtime:
+                duration = this.worked_hours - report._theoretical_hours(
+                    this.employee_id, date
+                )
+                self.env["hr.attendance.overtime"].create(
+                    {
+                        "employee_id": this.employee_id.id,
+                        "date": date,
+                        "duration": duration,
+                        "duration_real": duration,
+                    }
+                )
+            else:
+                overtime.write(
+                    {
+                        "duration": overtime.duration + this.worked_hours,
+                        "duration_real": overtime.duration + this.worked_hours,
+                    }
+                )
+
+    def _overtime_from_theoretical_time(self):
+        """
+        Write missing overtime records based on hr_attendance_theoretical_time_report
+        Generate records up until yesterday to avoid interfering with today's time recording
+        """
+        self.env.cr.execute(
+            """
+            select
+                r.employee_id, r.date, sum(r.theoretical_hours), sum(r.worked_hours),
+                sum(r.difference)
+            from hr_attendance_theoretical_time_report r
+            left join hr_attendance_overtime o
+                on r.employee_id=o.employee_id and r.date=o.date and not o.adjustment
+            where o.id is null and r.date < now()::date
+            group by r.employee_id, r.date
+            """
+        )
+        create_list = []
+        report = self.env["hr.attendance.theoretical.time.report"]
+        for (
+            employee_id,
+            date,
+            theoretical_hours,
+            worked_hours,
+            difference,
+        ) in self.env.cr.fetchall():
+            employee = self.env["hr.employee"].browse(employee_id)
+            duration = (
+                difference
+                if theoretical_hours > 0
```

Comment:
This fails in my setup.

```
File "/mnt/extra-addons/verdigado_attendance/models/hr_attendance.py", line 80, in _overtime_from_theoretical_time
     if theoretical_hours > 0
Exception

The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 654, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 301, in _handle_exception
     raise exception.with_traceback(None) from new_cause
TypeError: '>' not supported between instances of 'NoneType' and 'int'
```

This seems to happen with all lines where `hr_attendance_theoretical_time_report.theoretical_time` is 0.


I've worked around for testing with the following code before the duration calculation. This seems to work now.
```
if theoretical_hours is None:
                theoretical_hours = 0
```

(I just wonder, why 0 is NoneType but...)

Reply:
PS: ich finde [module_auto_update](https://github.com/OCA/server-tools/tree/15.0/module_auto_update) recht hilfreich sobald es um viele patches und Module geht - das speichert pro Modul eine checksum des codes und kann dann nur die geanderten Module updaten, das lasst sich auch recht einfach automatisieren.

Reply:
PPS: habe in der letzten version ein coalesce() zugefugt, damit sollte das auch fur Benutzer ohne Arbeitskalender fur den entsprechenden Zeitraum gehen

Reply:
Mit den Echt-Daten liefert die Berechnung teils völlig falsche Ergebnisse. Ich kann das bisher nur unvollständig beschreiben. Nur ein paar Stichpunkte. Wir sollten uns das vermutlich auf der selben Daten-Basis gemeinsam anschauen.

Mein aktueller Wissensstand: 

* Odoo-Overtime-Berechnung muss aktiv sein, damit das ganze überhaupt Sinn macht. Deshalb sollte auch keine Daten vor `res_company.overtime_start_date` durch Migration oder Cron angelegt werden.
* Beim Deaktivieren der Odoo-Overtime-Berechnung wird die Tabelle `hr_attendance_overtime` ja geleert. Beim Aktivieren erfolgt die Neuberechnung der Overtimes. Wir müssten dann noch mal `_overtime_from_theoretical_time()` ausführen, oder?
* Der Code required das OCA-Modul hr_attendance_report_theoretical_time. Für uns ist das ok.
* Nach der letzten Änderung stimmt die "duration" noch weniger. Hier steht jetzt der  Wert der `worked_hours` ohne Abzug der `theoretical_hours`, was zu einer riesigen Überstundenzahl führt. "Irgendwas" stimmt da nicht. Der Code zieht das ja theoretisch ab. Ich kann es aktuell nicht erklären.

Reply:
ich sehe dass wir mit dem heutigen Ansatz (=nehme die Ergebnisse der theoretischen Stunden) wohl nicht weiterkommen werden, deshalb verbringe ich den heutigen Tag damit das auf Basis direkt der Arbeitskalender zu berechnen. Bei der Gelegenheit ziehe ich das auch gleich in ein eigenes Modul um

Reply:
Ok. Dann wäre das Modul auch ohne `hr_attendance_report_theoretical_time` nutzbar? Das wäre sicher gut und universeller.